### PR TITLE
adding prompt-as-a-feature variables, modifying inference and train f…

### PR DESCRIPTION
**Pull request by AmeyHengle**
…unctions

### CODE REVIEW

File: dialogy/plugins/text/classification/xlmr.py [hunk 1]
```diff
@@ -148,8 +161,9 @@ def valid_labelencoder(self) -> bool:
         return hasattr(self.labelencoder, "classes_")
 
     def inference(
-        self, texts: Optional[List[str]], state: Optional[str] = None
+        self, texts: Optional[List[str]], state: Optional[str] = None, lang: Optional[str] = None,
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:164]: Please add `lang` in docstring

File: dialogy/plugins/text/classification/xlmr.py [hunk 2]
```diff
@@ -148,8 +161,9 @@ def valid_labelencoder(self) -> bool:
...
     ) -> List[Intent]:
+
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:166]:
```suggestion
```

File: dialogy/plugins/text/classification/xlmr.py [hunk 3]
```diff
@@ -42,8 +44,13 @@ def __init__(
         data_column: str = const.DATA,
         label_column: str = const.LABELS,
         state_column: str = const.STATE,
+        lang_column: str = const.LANG,
         args_map: Optional[Dict[str, Any]] = None,
         skip_labels: Optional[List[str]] = None,
+        prompts_map: Dict[Any, Any] = {},
+        use_prompt: bool = False,
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:51]: Why can't we make use of `use_state` here by assuming that `prompt` is part of the state?

File: dialogy/plugins/text/classification/xlmr.py [hunk 4]
```diff
@@ -176,13 +192,36 @@ def inference(
             raise ValueError(
                 f"Plugin {self.__class__.__name__} requires state to be passed to the model."
             )
-        elif self.use_state and state:
-            texts = [f"{text} <s> {state} </s>" for text in texts]
+
+        if self.use_prompt and not state:
+            raise ValueError(
+                f"In order to use prompts as feature, Plugin {self.__class__.__name__} is requires state to be passed to the model."
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:198]:
```suggestion
                f"In order to use prompts as feature, plugin {self.__class__.__name__} requires state to be passed to the model."
```

File: dialogy/plugins/text/classification/xlmr.py [hunk 5]
```diff
@@ -176,13 +192,36 @@ def inference(
...
+            )
+
+        if self.use_prompt and not lang:
+            raise ValueError(
+                f"In order to use prompts as feature, Plugin {self.__class__.__name__} is requires lang to be passed to the model."
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:203]:
```suggestion
                f"In order to use prompts as feature, plugin {self.__class__.__name__} requires lang to be passed to the model."
```

File: dialogy/plugins/text/classification/xlmr.py [hunk 6]
```diff
@@ -176,13 +192,36 @@ def inference(
...
+            )
+
+        if self.use_prompt and state:
+            texts[0] += "<s> " + self.lookup_prompt(lang=lang, state=state)[0] + " </s>"
+            if self.debug:
+                logger.debug(f"Classification input (Prompt as a feature enabled):\n{texts}")
+                logger.debug(texts)
+                # logger.debug(self.lookup_prompt(lang, state)[0])
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:211]:
```suggestion
```

File: dialogy/plugins/text/classification/xlmr.py [hunk 7]
```diff
@@ -176,13 +192,36 @@ def inference(
...
+            )
+
+        if self.use_prompt and not lang:
+            raise ValueError(
+                f"In order to use prompts as feature, Plugin {self.__class__.__name__} is requires lang to be passed to the model."
+            )
+
+        if self.use_prompt and state:
+            texts[0] += "<s> " + self.lookup_prompt(lang=lang, state=state)[0] + " </s>"
+            if self.debug:
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:208]: I don't think you should be needing an if construct here if logging level is properly configured while running the framework.

File: dialogy/plugins/text/classification/xlmr.py [hunk 8]
```diff
@@ -176,13 +192,36 @@ def inference(
...
+                logger.debug(f"Classification input (Prompt as a feature enabled):\n{texts}")
+                logger.debug(texts)
+                # logger.debug(self.lookup_prompt(lang, state)[0])
+
+        if self.use_state and state:
+            texts[0] += "<s> " + state + " </s>"
+            if self.debug:
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:215]: Same about the `if` construct here

File: dialogy/plugins/text/classification/xlmr.py [hunk 9]
```diff
@@ -257,14 +303,58 @@ def train(self, training_data: pd.DataFrame) -> None:
...
             training_data[const.LABELS]
         )
-        # Add state as an additonal field to text
+
+        # Append state to text
         if self.use_state:
-            training_data[const.TEXT] += "<s> " + training_data["state"] + " </s>"
+            # logger.debug(training_data[const.STATE])
+            training_data[const.TEXT] += "<s> " + training_data[self.state_column] + " </s>"
+
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:311]: Please move out the addition of state / prompt to the texts in an independent function and reuse that function across `train` and `inference` methods.

File: dialogy/plugins/text/classification/xlmr.py [hunk 10]
```diff
@@ -257,14 +303,58 @@ def train(self, training_data: pd.DataFrame) -> None:
...
+        # Append prompt to text
+        if self.use_prompt:
+            logger.debug("Adding prompts to input text")
+
+            if not self.train_using_all_prompts:
+                for i in tqdm(range(training_data.shape[0]), desc="progress bar:"):
+                    _lang = training_data.iloc[i][self.lang_column]
+                    _state = training_data.iloc[i][self.state_column]
+                    _prompt =  self.lookup_prompt(lang=_lang, state=_state)[0]
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:320]: Here, we are employing a strategy to choose a single prompt out of all available prompts. This should live inside a function and reused across `train` and `inference` methods.

File: dialogy/plugins/text/classification/xlmr.py [hunk 11]
```diff
@@ -257,14 +303,58 @@ def train(self, training_data: pd.DataFrame) -> None:
...
+                        # training_data_ = training_data_.append(_row, ignore_index=True)
+                        training_data_.reset_index(drop=True)
+            
+                logger.debug(f"Training dataset size (original): {training_data.shape}")
+                logger.debug(f"Training dataset size (after adding prompts): {training_data_.shape}")
+
+                training_data = training_data_
+                del training_data_
+
+        training_data.to_csv("/tmp/xlmr-data-train.csv",index=False)
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:351]: Why do we need to store this to a `tmp` folder?

File: dialogy/plugins/text/classification/xlmr.py [hunk 12]
```diff
@@ -285,6 +375,18 @@ def save(self) -> None:
             writer=pickle.dump,
         )
 
+    def lookup_prompt(self, lang: Optional[str], state: Optional[str], return_all: bool = False) -> List[str]:
+        """
+        Same as get_prompt() method, but built for faster lookup to reduce latency during inference. 
+        """
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:381]: Please add description of params and return value to the docstring.

File: dialogy/plugins/text/classification/xlmr.py [hunk 13]
```diff
@@ -285,6 +375,18 @@ def save(self) -> None:
...
+        try:
+            return random.sample(self.prompts_map[lang][state], 1)
+        except Exception as e:
+            if self.debug:
```
Comment (dakshvar22) [inline dialogy/plugins/text/classification/xlmr.py:385]: `if` construct shouldn't be needed.

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Haven't looked at the tests yet, but made suggestions to improve the reusability of code.

**Comment by AmeyHengle** [comment]
This PR is redundant. Here, current_state is used as the primary key for fetching prompts. 
Closing this, I have pushed a new PR which has changes in accordance with our final discussion. (primary key is nls_label)
@ltbringer

**Pull request closed without merging**